Use unified memory for conditions

makortel opened the pull request:
This PR experiments using unified memory for conditions. It adds a helper class `CUDAESManaged` to simplify calling the `cudaMemAdvise(..., cudaMemAdviseSetReadMostly, 0)` and `cudaMemPrefetchAsync(...)` to all allocated buffers.

For the CPE and the cabling map it also experiments passing a struct of GPU pointers to the kernel instead of a GPU pointer to a struct of GPU pointers.

It also adds `CUDAManagedAllocator` and `CUDAManagedVector<T>` because I thought first that I'd use them, but in the end didn't.

I have not done a detailed performance evaluation wrt. the current state.

fwyzard commented:
:construction: Validation running at workergpu15:/mnt/home/fwyzard/patatrack/validation/run.DImywVMW8A ...

fwyzard commented:
:construction: Validation running at vinavx2.cern.ch:/data/user/fwyzard/patatrack/validation/run.SNc9zhL3Wc ...

fwyzard commented:
@makortel @VinInn can you double check if the merge looks good ?

fwyzard commented:
## Reference
Throughput over 1000 events:
  - mean: 629 ± 42 ev/s
  - best: 644 ± 31 ev/s

Top 10 contribution to GPU usage:
```
  Time(%)      Time     Calls       Avg       Min       Max  Name
   51.96%  855.55ms      1200  712.96us  200.22us  4.5703ms  gpuClustering::findClus(...)
   21.42%  352.71ms      1200  293.92us  100.48us  720.02us  gpuPixelDoublets::getDoubletsFromHisto(...)
    8.07%  132.86ms      1200  110.72us  27.552us  385.02us  kernel_connect(...)
    7.78%  128.09ms      1200  106.74us  19.744us  248.41us  kernel_find_ntuplets(...)
    3.65%   60.16ms      1200  50.131us  39.583us  64.383us  gpuPixelRecHits::getHits(...)
    2.63%   43.26ms      6028  7.1760us  1.2160us  519.58us  [CUDA memcpy HtoD]
    1.29%   21.26ms      1200  17.712us  5.4720us  56.575us  kernel_checkOverflows(...)
    0.68%   11.14ms      1200  9.2860us  5.9200us  34.079us  pixelgpudetails::RawToDigi_kernel(...)
    0.56%    9.18ms      6014  1.5250us  1.1200us  18.079us  [CUDA memset]
    0.41%    6.71ms      1200  5.5910us  3.4880us  8.8960us  void cub::DeviceScanKernel<cub::DispatchScan<unsigned int*, unsigned int*, cub::Sum, cub::NullType, int>::PtxAgentScanPolic...
```

## Pull request #157,
Throughput over 1000 events:
  - mean: 613 ± 29 ev/s
  - best: 614 ± 28 ev/s

```
Top 10 contribution to GPU usage:
  Time(%)      Time     Calls       Avg       Min       Max  Name
   52.51%  851.08ms      1200  709.24us  217.28us  4.5735ms  gpuClustering::findClus(...)
   21.78%  352.96ms      1200  294.13us  97.758us  709.34us  gpuPixelDoublets::getDoubletsFromHisto(...)
    8.18%  132.61ms      1200  110.50us  26.432us  372.38us  kernel_connect(...)
    7.62%  123.46ms      1200  102.89us  23.455us  236.16us  kernel_find_ntuplets(...)
    2.93%   47.57ms      1200  39.644us  29.951us  59.328us  gpuPixelRecHits::getHits(...)
    2.54%   41.21ms      6014  6.8510us  1.2470us  45.248us  [CUDA memcpy HtoD]
    1.30%   21.11ms      1200  17.590us  6.5920us  55.231us  kernel_checkOverflows(...)
    0.58%    9.37ms      1200  7.8040us  4.1600us  26.175us  pixelgpudetails::RawToDigi_kernel(...)
    0.56%    9.05ms      6014  1.5040us  1.1200us  11.872us  [CUDA memset]
    0.40%    6.56ms      1200  5.4680us  3.4560us  8.7680us  void cub::DeviceScanKernel<cub::DispatchScan<unsigned int*, unsigned int*, cub::Sum, cub::NullType, int>::PtxAgentScanPolic...
```

fwyzard commented:
While the contribution of the individual kernels does not seem to change, there seems to be an overall degradation of performance.

makortel commented:
@fwyzard 

> feel free to squash away my "merge" commits, of even just rebase on top ofthe current HEAD.

Rebased on top of the current HEAD.

makortel commented:
@fwyzard 

> While the contribution of the individual kernels does not seem to change, there seems to be an overall degradation of performance.
> 
> I think it should be addressed - or at least understood - before merging.

My understanding is that some slowdown is expected, even with prefetching. But I agree it is a bit strange to see slowdown after the first events (where, despite of prefetching, I saw GPU page faults in the profile). 

One possibility is that currently I call the `cudaMemPrefetchAsync()` on each event (for each memory buffer of each ESProduct). Even if it doesn't have to actually prefetch the data after the first event, it likely still has to do something e.g. to check where the memory resides currently. Maybe some ultra-lightweight check could be added so that for the later events the call could be avoided.

makortel commented:
@fwyzard I reduced the amount of calls to `cudaMemPrefetchAsync()` with an `atomic<bool>`. Could you check if it has any effect on your test?

fwyzard commented:
I have repeated a comparison of the performance with the latest changes, measuring the throughput over 4000 pixel-only events, running on P100 GPUs:
  - 1 job, with 1 stream, 1 CPU thread, 1 GPU
  - 2 jobs, each with 1 stream, 1 CPU thread, 1 GPU
  - 1 job, with 4 streams. 4 CPU threads, 2 GPUs

The throughput for the three cases is:
 | original | #157 
-- | -- | -- | -- | --
1 x (1 CPU, 1 GPU)  |  238 ev/s |  ± 3 ev/s |  233 ev/s | ± 2 ev/s
2 x (1 CPU,  1 GPU)  |  476 ev/s |  ± 4 ev/s |  465 ev/s | ± 3 ev/s
1 x (4 CPUs, 2 GPUs)  |  521 ev/s |  ± 4 ev/s |  523 ev/s | ± 4 ev/s

So, for jobs running with a single CPU and a single GPU, we see a ~2% slowdown, while for jobs running with multiple CPUs and GPUs the performance are the same.

fwyzard commented:
Note that in all these cases there are no conditions updates after the first event, so I'm not sure what we are measuring, exactly...

makortel commented:
With the present state of the PR, after the first event(s) both the "original" and the "PR" essentially just check an `atomic<bool>` (in CPU) that they don't have to do anything. Beyond that, the only difference I can think of is (quoting from the description)
> For the CPE and the cabling map it also experiments passing a struct of GPU pointers to the kernel instead of a GPU pointer to a struct of GPU pointers.

makortel commented on HeterogeneousCore/CUDACore/src/CUDAESManaged.cc:
```diff
@@ -0,0 +1,30 @@
+  // matter if more than one thread/edm stream issues the prefetches
+  // as long as most of the prefetches are avoided.
+  if(prefetched_.load())
+    return;
```
Looking through the code again I noticed that here we/I have a single flag to control behaviour for all devices. I think this should be changed to either
1. the `atomic<bool>` flag should be changed to per-device flags, or
2. issue the prefetch for all devices below

In principle the latter (2) could make more sense as we expect all devices to run all code (and thus need all GPU conditions) anyway. But async prefetch needs also a CUDA stream, so it needs a bit more thought. Basically it would mean that the CUDAESManaged would have additional CUDA streams (one per device), and then use CUDA events to "synchronize" with the argument stream (otherwise there would be no gain from async). Then the `atomic<bool>` flag would actually be replaced with `cudaEventQuery()` for the CUDA event (of a device). I would naively imagine `atomic<bool>` to be faster than `cudaEventQuery()`, which would favor option 1.

makortel commented:
Rebased on top of 10_2_**5**, and
* Fixed the boolean flag issue (has now an atomic per device)
* Went back to GPU struct of pointers to GPU

I'm not planning any further updates to this PR.

makortel commented:
Rebased on top of CMSSW_10_4_0_pre4_Patatrack. Note that so far I've only compiled it, and didn't test running.

makortel commented:
Rebased on top of CMSSW_10_4_0_pre4_Patatrack to fix conflicts. Note that so far I've only compiled it, and didn't test running.
